apps/builder/app/builder/features/breakpoints/calc-canvas-width.ts

```diff
@@ -0,0 +1,129 @@
+import { compareMedia } from "@webstudio-is/css-engine";
+import type { Breakpoint } from "@webstudio-is/sdk";
+import {
+  groupBreakpoints,
+  isBaseBreakpoint,
+  minCanvasWidth,
+} from "~/shared/breakpoints";
+
+const defaultWidth = 320;
+
+// We want to know if user resized the canvas to a custom value.
+const isCustomCanvasWidth = (
+  breakpoints: Array<Breakpoint>,
+  selectedBreakpoint: Breakpoint,
+  canvasWidth?: number
+) => {
+  if (canvasWidth === undefined) {
+    return false;
+  }
+  const hasMinWidth = breakpoints.some(
+    (breakpoint) => breakpoint.minWidth !== undefined
+  );
+
+  // Only in case of base breakpoint and whithout min-width breakpoint, we ignore the custom width,
+  // because we want it to go full viewport width
+  if (isBaseBreakpoint(selectedBreakpoint) && hasMinWidth === false) {
+    return false;
+  }
+
+  for (const breakpoint of breakpoints) {
+    if (
+      breakpoint.minWidth !== undefined &&
+      canvasWidth === breakpoint.minWidth
+    ) {
+      return false;
+    }
+    if (
+      breakpoint.maxWidth !== undefined &&
+      canvasWidth === breakpoint.maxWidth + 1
+    ) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Find a canvas width that is within the selected breakpoint's range, but is at it's minimum.
+// Goal is to allow user to consistently know the direction they want to resize to after they switched to a breakpoint.
+// In this case they will want to always increase the width after switching.
+export const calcCanvasWidth = ({
+  breakpoints,
+  selectedBreakpoint,
+  workspaceWidth,
+  canvasWidth,
+}: {
+  breakpoints: Array<Breakpoint>;
+  selectedBreakpoint: Breakpoint;
+  workspaceWidth: number;
+  canvasWidth?: number;
+}) => {
+  // When user has resized the canvas to a custom value, we want to keep it.
+  if (isCustomCanvasWidth(breakpoints, selectedBreakpoint, canvasWidth)) {
+    return canvasWidth;
+  }
+
+  // Finding the canvas width when user selects base breakpoint is a bit more complicated.
+  // We want to find the lowest possible size that is bigger than all max breakpoints and smaller than all min breakpoints.
+  // Note: it is still possible to get intersecting min and max breakpoints.
+  if (isBaseBreakpoint(selectedBreakpoint)) {
+    // Base is the only breakpoint
+    if (breakpoints.length === 1) {
+      return workspaceWidth;
+    }
+
+    const grouped = groupBreakpoints(breakpoints).filter(
+      ({ minWidth, maxWidth }) => {
+        if (minWidth && minWidth < workspaceWidth) {
```

Comment:
Is it possible that we are on base breakpoint and minWidth < workspaceWidth?

Reply:
I mean greater ;)

Reply:
yes its possible, why not, you can have a tiny workspace and a large min-width, this is why we have scaling , so you can test on a large breakpoint without having a large screen